text/0009-client-api.md

```diff
@@ -0,0 +1,154 @@
+| /query  | "state" | { "block_hash": \<HEX STRING\> or null, "key": \<HEX STRING\>, "path": ["key1", "key2", etc], "type": \<STRING\> } | CLValue |
+| /query  | "peers" | none                             | API version + list of node IDs (not validator IDs) and endpoints |
+| /query  | "all"   | none                             | API version + list of peers + the latest block                   |
+
+Note: For /block "get", if no params are passed, the latest block is returned.
+
+Note: For /query "state", if "block_hash" is null, the latest block will be used.  The "type" must be
+one of "hash", "uref", or "address".  The returned value on success will be `CLValue`, serialized
+using `ToBytes`, then hex-encoded.
+
+
+### Event stream
+
+The path for the event stream is `/events`.  It will support a query string with the following
+optional components:
+
+* `subscription=[comma-separated list of event types]` where the available event types are
+    * `block_finalized`
+    * `block_added`
+    * `deploy_added`
+    * `deploy_requeued`
+    * `deploy_discarded`
+    * `deploy_finalized`
+    * `deploy_processed`
+    * `deploy_orphaned`
+* alternatively, `subscription=all` implies subscribing to all available event types
+* `accounts=[comma-separated list of hex-encoded accounts]` to filter by the listed accounts
+* `deploy_hashes=[comma-separated list of hex-encoded deploy hashes]` to filter by the listed deploys
+
+The events corresponding to these subscriptions are as close as possible to those currently defined
+in [info.proto](https://github.com/CasperLabs/CasperLabs/blob/cd2e80286d4172c9d2e73a2c1f1271b2961a7527/protobuf/io/casperlabs/casper/consensus/info.proto#L91-L161).
+
+The first message returned will be the API version.  All subsequent ones will be JSON-encoded events.
+
+### Use case
+
+It's anticipated that the Rust client would be able to send a new deploy using the /deploy "put" RPC, while also having
+subscribed to events filtered on that deploy's ID.
+
+In this way, the client's `put-deploy` and `send-deploy` subcommands could accept arguments specifying that the client
+wait until the deploy has transitioned to certain state.  Internally, the client will consume events (possibly reporting
+each) until the requested deploy state has been reached, at which point the client can disconnect.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+* Clients will not be able to receive as rich information in response to requests for information as they currently do.
+
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* Plain JSON requests and responses via a REST API is a viable alternative to JSON-RPCs.  There
+appears to be little difference in usability or implementation effort, although other projects in
+the blockchain space use JSON-RPC, so perhaps there would be some slight benefit for us to follow
+suit.
```

Comment:
I agree with this statement; I think the primary reason to use json-rpc over simple rest is to adhere to expected convention within the space.